File: src/grid.cpp
#include "grid.hpp"
#include <array>
#include <cmath>
#include <numeric>  // for std::inner_product
#include <omp.h>
#include <thread>
#include <vector>

// NOLINTBEGIN(*internal-linkage)
#ifdef NDEBUG
namespace {
#endif
/**
 * @brief mock numpy.linspace(0)
 *
 * @param min minimal value of the data
 * @param max maximal value of the data
 * @param binnum number of intervals (not the number of data point)
 * @param withRightBound whether include the righter bound
 * @return an std::vector<double> for the bin edges
 */
auto linspace(double min,
              double max,
              int    binnum,
              bool   withRightBound = true) noexcept
{
    auto delta = (max - min) / binnum;

    auto effBinNum{binnum + 1};
    if (not withRightBound)
    {
        --effBinNum;
    }

    std::vector<double> bin_edges(effBinNum, 0);
    for (int i = 0; i < effBinNum; ++i)
    {
        bin_edges[i] = min + (delta * i);
    }
    return bin_edges;
};
#ifdef NDEBUG
}  // namespace
#endif
// NOLINTEND(*internal-linkage)

double GridPoint::radius() const
{
    // get the radius of the grid point
    auto r2 =
        std::inner_product(m_pos.begin(), m_pos.end(), m_pos.begin(), 0.0);
    return std::sqrt(r2);
}

// pass by value, since array<double, 3> is cheap to copy
double GridPoint::distance_from(std::array<double, 3> coordinate) const
{
    for (int i = 0; i < 3; ++i)  // calculate \Delta\vec{r}
    {
        // NOLINTBEGIN(*array-index)
        coordinate[i] -= m_pos[i];
        // NOLINTEND(*array-index)
    }

    // get the norm through inner product
    auto r2 = std::inner_product(coordinate.begin(), coordinate.end(),
                                 coordinate.begin(), 0.0);
    return std::sqrt(r2);
}

double GridPoint::radial_comp(std::array<double, 3> force) const
{
    const double r = radius();  // radial of the GridPoint

    if (r == 0)  // case where the point is at the origin point
    {
        // in this case, the result is phyically unclear (or undefiend)
        // return the norm of the force vector
        auto f2 =
            std::inner_product(force.begin(), force.end(), force.begin(), 0.0);
        return std::sqrt(f2);
    }

    // scalar product between the force and the unit radial vector when r!=0
    return dot_with(force) / r;
}

auto GridPoint::accR_from(const std::vector<double>&                masses,
                          const std::vector<std::array<double, 3>>& coordinates,
                          double G) const -> double
{
    // to restore the Ftot= G \sum{ m delta(r) / r^3 }
    std::array<double, 3> accSum{0, 0, 0};

    // NOLINTBEGIN(*array-index)
    // iteration to get the sum without times G
    for (int i = 0; i < static_cast<int>(masses.size()); ++i)
    {
        static double d{};       // for r^3
        static double scalar{};  // for m/r^3
        d      = distance_from(coordinates[i]);
        scalar = masses[i] / (d * d * d);
        accSum[0] += scalar * (coordinates[i][0] - m_pos[0]);
        accSum[1] += scalar * (coordinates[i][1] - m_pos[1]);
        accSum[2] += scalar * (coordinates[i][2] - m_pos[2]);
    }
    // times G
    accSum[0] *= G;
    accSum[1] *= G;
    accSum[2] *= G;
    // NOLINTEND(*array-index)

    return radial_comp(accSum);
}

PolarGrid::PolarGrid(const PolarGridPara& para)
{
    const auto numOfPoint = (para.rbin + 1) * para.phibin;
    m_points.reserve(numOfPoint);  // enlarge the capacity of the list

    // get the r&phi bin edges
    m_phibinEdges = linspace(0.0, 2 * M_PI, para.phibin, false);  // 0 to 2pi
    if (para.type == RbinType::linear)                            // linear bins
    {
        m_rbinEdges = linspace(para.rmin, para.rmax, para.rbin, true);
    }
    else  // logarithmic bins
    {
        m_rbinEdges = linspace(std::log10(para.rmin), std::log10(para.rmax),
                               para.rbin, true);
        std::transform(m_rbinEdges.begin(), m_rbinEdges.end(),
                       m_rbinEdges.begin(),
                       [](double r) { return std::pow(10, r); });
    }

    // create the target filed point
    for (auto r : m_rbinEdges)
    {
        for (auto phi : m_phibinEdges)
        {
            m_points.emplace_back(r * std::cos(phi), r * std::sin(phi), 0);
        }
    }
};

auto PolarGrid::cal_accR_from(
    const std::vector<double>&                masses,
    const std::vector<std::array<double, 3>>& coordinates,
    int numThread) const -> std::vector<double>
{
    ( void )numThread;
    std::vector<double> accR(m_points.size(), 0);
    // clang-format off
#pragma omp parallel for num_threads(numThread) default(none) shared(accR, masses, coordinates)
    for (int i = 0; i < static_cast<int>(m_points.size()); ++i) // i must be int for openmp to work
    {
        accR[i] = m_points[i].accR_from(masses, coordinates);  // with default G
    }
    // clang-format on
    return accR;
}

auto PolarGrid::rs() const -> std::vector<double>
{
    std::vector<double> rs;
    rs.reserve(m_points.size());
    // create the target filed point
    for (auto r : m_rbinEdges)
    {
        for ([[maybe_unused]] auto phi : m_phibinEdges)
        {
            rs.emplace_back(r);
        }
    }
    return rs;
}

auto PolarGrid::phis() const -> std::vector<double>
{
    std::vector<double> phis;
    phis.reserve(m_points.size());
    // create the target filed point
    for ([[maybe_unused]] auto r : m_rbinEdges)
    {
        for (auto phi : m_phibinEdges)
        {
            phis.emplace_back(phi);
        }
    }
    return phis;
}

File: rust/src/grid.rs
/// Polar grid of field points, used to compute radial gravitational
/// accelerations from a particle set via Newton summation.
///
/// The call order is:
/// 1. [`PolarGrid::new`] to create the grid from parameters
/// 2. [`PolarGrid::cal_acc_r_from`] to compute accelerations in parallel
/// 3. [`PolarGrid::rs`] / [`PolarGrid::phis`] to export flattened coordinates

use rayon::prelude::*;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/// Numerical gravitational constant in Gadget-4 internal units
/// (kpc, km/s, 1e10 M☉).
const G: f64 = 43007.1;

/// π to full precision.
const PI: f64 = 3.14159265358979323846264338327950288;

// ---------------------------------------------------------------------------
// Radial bin type
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RbinType {
    Log,
    Linear,
}

// ---------------------------------------------------------------------------
// Grid parameters (aggregate input)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy)]
pub struct PolarGridPara {
    pub rmin: f64,
    pub rmax: f64,
    pub rbin: i32,
    pub phibin: i32,
    pub rtype: RbinType,
}

// ---------------------------------------------------------------------------
// Single grid point
// ---------------------------------------------------------------------------

/// A single field point where the radial acceleration is evaluated.
#[derive(Debug, Clone, Copy)]
pub struct GridPoint {
    x: f64,
    y: f64,
    z: f64,
}

impl GridPoint {
    #[inline(always)]
    pub const fn new(x: f64, y: f64, z: f64) -> Self {
        Self { x, y, z }
    }

    /// Cylindrical radius √(x² + y²).
    #[inline(always)]
    pub fn cyl_radius(&self) -> f64 {
        (self.x * self.x + self.y * self.y).sqrt()
    }

    /// Radial component (cylindrical) of a force vector acting on this point.
    ///
    /// When the point lies on the z-axis (`R == 0`) the result is the
    /// magnitude of the force projected onto the xy-plane.
    #[inline(always)]
    fn cyl_radial_comp(&self, force: &[f64; 3]) -> f64 {
        let r = self.cyl_radius();
        if r == 0.0 {
            // On-axis: return norm of xy-plane force
            (force[0] * force[0] + force[1] * force[1]).sqrt()
        } else {
            // Off-axis: dot(force, unit radial vector)
            (force[0] * self.x + force[1] * self.y) / r
        }
    }

    /// Compute the cylindrical radial acceleration at this point produced by
    /// an ensemble of particles.
    ///
    /// # Safety
    ///
    /// - `masses` and `coordinates` must have the same length.
    /// - `coordinates` is a flat Vec<f64> with stride 3 per particle.
    ///
    /// The caller guarantees both invariants, so we deliberately skip all
    /// bounds checking inside the critical loop.
    #[inline(always)]
    unsafe fn acc_r_from_flat_unchecked(
        &self,
        masses: &[f64],
        coord_flat: &[f64],
    ) -> f64 {
        // Stack-allocated accumulator
        let mut sx = 0.0_f64;
        let mut sy = 0.0_f64;
        let mut sz = 0.0_f64;

        let n = masses.len();
        let mp = masses.as_ptr();
        let cp = coord_flat.as_ptr();

        for i in 0..n {
            // SAFETY: caller guarantees length match
            let m = *mp.add(i);
            let cx = *cp.add(3 * i);
            let cy = *cp.add(3 * i + 1);
            let cz = *cp.add(3 * i + 2);

            let dx = cx - self.x;
            let dy = cy - self.y;
            let dz = cz - self.z;

            let r2 = dx * dx + dy * dy + dz * dz;
            let inv_r3 = 1.0 / (r2 * r2.sqrt()); // 1 / r³

            let scalar = m * inv_r3; // m / r³
            sx += scalar * dx;
            sy += scalar * dy;
            sz += scalar * dz;
        }

        // Multiply by G once, outside the loop.
        let acc = [sx * G, sy * G, sz * G];
        self.cyl_radial_comp(&acc)
    }
}

// ---------------------------------------------------------------------------
// Polar grid
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct PolarGrid {
    points: Vec<GridPoint>,
    rbin_edges: Vec<f64>,
    phibin_edges: Vec<f64>,
    rbin: i32,
    phibin: i32,
}

/// Create a linearly-spaced vector of bin edges.
fn linspace(min: f64, max: f64, binnum: i32, with_right_bound: bool) -> Vec<f64> {
    let delta = (max - min) / f64::from(binnum);
    let n = if with_right_bound { binnum + 1 } else { binnum } as usize;
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        out.push(min + delta * i as f64);
    }
    out
}

impl PolarGrid {
    /// Build the polar grid from the given parameters.
    pub fn new(para: &PolarGridPara) -> Self {
        let rbin = para.rbin;
        let phibin = para.phibin;

        let phibin_edges = linspace(0.0, 2.0 * PI, phibin, false);

        let rbin_edges = match para.rtype {
            RbinType::Linear => linspace(para.rmin, para.rmax, rbin, true),
            RbinType::Log => {
                let raw = linspace(para.rmin.log10(), para.rmax.log10(), rbin, true);
                raw.into_iter().map(|v| 10.0_f64.powf(v)).collect()
            }
        };

        let npoints = (rbin + 1) as usize * phibin as usize;
        let mut points = Vec::with_capacity(npoints);
        for &r in &rbin_edges {
            for &phi in &phibin_edges {
                points.push(GridPoint::new(r * phi.cos(), r * phi.sin(), 0.0));
            }
        }

        Self {
            points,
            rbin_edges,
            phibin_edges,
            rbin,
            phibin,
        }
    }

    /// Compute the cylindrical radial acceleration at every grid point from
    /// the given particle set.
    ///
    /// Accepts flat coordinates for zero-copy interop with HDF5 output.
    /// Each grid point is processed independently in parallel using Rayon.
    pub fn cal_acc_r_from_flat(
        &self,
        masses: &[f64],
        coord_flat: &[f64],
        pool: &rayon::ThreadPool,
    ) -> Vec<f64> {
        let points = &self.points;
        let mut acc_r = vec![0.0_f64; points.len()];

        pool.install(|| {
            acc_r.par_iter_mut().enumerate().for_each(|(i, a)| {
                // SAFETY: masses.len() * 3 == coord_flat.len(), caller guarantees
                *a = unsafe { points[i].acc_r_from_flat_unchecked(masses, coord_flat) };
            });
        });

        acc_r
    }

    /// Create a Rayon thread pool with the specified number of threads.
    pub fn create_pool(num_thread: usize) -> rayon::ThreadPool {
        rayon::ThreadPoolBuilder::new()
            .num_threads(num_thread)
            .build()
            .expect("Failed to build Rayon thread pool")
    }

    /// Flattened radial coordinates of all grid points.
    pub fn rs(&self) -> Vec<f64> {
        let mut out = Vec::with_capacity(self.points.len());
        for &r in &self.rbin_edges {
            for _ in 0..self.phibin as usize {
                out.push(r);
            }
        }
        out
    }

    /// Flattened azimuthal angles of all grid points.
    pub fn phis(&self) -> Vec<f64> {
        let mut out = Vec::with_capacity(self.points.len());
        for _ in &self.rbin_edges {
            for &phi in &self.phibin_edges {
                out.push(phi);
            }
        }
        out
    }

    pub fn rbin_edges(&self) -> &[f64] {
        &self.rbin_edges
    }

    pub fn phibin_edges(&self) -> &[f64] {
        &self.phibin_edges
    }
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linspace_inclusive() {
        let v = linspace(0.0, 1.0, 4, true);
        assert_eq!(v.len(), 5);
        let expected = [0.0, 0.25, 0.5, 0.75, 1.0];
        for (a, b) in v.iter().zip(expected.iter()) {
            assert!((a - b).abs() < 1e-15);
        }
    }

    #[test]
    fn test_linspace_exclusive() {
        let v = linspace(0.0, 2.0 * PI, 16, false);
        assert_eq!(v.len(), 16);
        for (i, &val) in v.iter().enumerate() {
            let expected = (2.0 * PI / 16.0) * i as f64;
            assert!((val - expected).abs() < 1e-15);
        }
    }

    #[test]
    fn test_gridpoint_cyl_radius() {
        let gp = GridPoint::new(3.0, 4.0, 0.0);
        assert!((gp.cyl_radius() - 5.0).abs() < 1e-15);
    }

    #[test]
    fn test_single_particle_acc_r_flat() {
        let gp = GridPoint::new(1.0, 0.0, 0.0);
        let masses = vec![1.0];
        let coords = vec![1.0, 0.0, 0.0];
        let _acc = unsafe { gp.acc_r_from_flat_unchecked(&masses, &coords) };
    }

    #[test]
    fn test_two_particles_symmetric_flat() {
        let gp = GridPoint::new(0.0, 0.0, 0.0);
        let masses = vec![1.0, 1.0];
        let coords = vec![-1.0, 0.0, 0.0, 1.0, 0.0, 0.0];
        let acc = unsafe { gp.acc_r_from_flat_unchecked(&masses, &coords) };
        assert!(acc.abs() < 1e-10);
    }

    #[test]
    fn test_polar_grid_rs_phis_length() {
        let para = PolarGridPara {
            rmin: 0.1, rmax: 10.0, rbin: 5, phibin: 4, rtype: RbinType::Log,
        };
        let grid = PolarGrid::new(&para);
        assert_eq!(grid.rs().len(), (5 + 1) * 4);
        assert_eq!(grid.phis().len(), (5 + 1) * 4);
    }

    #[test]
    fn test_polar_grid_cal_acc_r_flat() {
        let para = PolarGridPara {
            rmin: 1.0, rmax: 2.0, rbin: 2, phibin: 2, rtype: RbinType::Linear,
        };
        let grid = PolarGrid::new(&para);
        let masses = vec![1e10];
        let coords = vec![0.0, 0.0, 0.0];
        let pool = PolarGrid::create_pool(2);
        let acc = grid.cal_acc_r_from_flat(&masses, &coords, &pool);
        assert_eq!(acc.len(), (2 + 1) * 2);
    }
}